### ISSUE TITLE
Audio cracks/gaps on Ventura

### ISSUE THREAD

**Issue by luismrguimaraes** [ExistentialAudio/BlackHole#824]
### Please Confirm

- [X] I have read the **[FAQ](https://github.com/ExistentialAudio/BlackHole#faq) and [Wiki](https://github.com/ExistentialAudio/BlackHole/wiki)** where most common issues can be resolved
- [X] I have searched **[Discussions](https://github.com/ExistentialAudio/BlackHole/discussions)** to see if the same question has already been asked
- [X] This is a **bug** and not a question about audio routing or configuration, which should be posted in [Discussions](https://github.com/ExistentialAudio/BlackHole/discussions)

### macOS Version

macOS 13 Ventura

### BlackHole Build(s) Affected

- [X] 2 channel
- [ ] 16 channel
- [ ] 64 channel
- [ ] other/custom build

### Describe the bug

I keep getting cracks/gaps while using the computer normally. I've found three specific instances where I can easily reproduce this. I use an app called [Simplenote](https://simplenote.com/) and whenever a note is saved, there is a crack/gap in the audio that is playing. Also, when using the `cmd+F` search feature in Firefox, I also get cracks whenever my prompt is not found. Finally, when screenshotting with `cmd+shift+4` it also happens.
I'm routing the audio like this: Blackhole -> DSP software -> Headphones. 

So far, I've tested playback in Reaper and the JUCE AudioPluginHost and also tested recording in Audacity. The cracks don't happen when I send the audio directly to the headphones, that is why I believe the issue is with blackhole, but I'm available to do further testing.

### Reproduction Steps

1. Set audio output device to blackhole
2. Set up a way to record or play the audio coming from blackhole
3. Perform any of the "triggers" described above

### Expected Behavior

No cracks.

### Screenshots

My aggregate device:
<img width="1207" alt="image" src="https://github.com/user-attachments/assets/180db556-9963-41cd-b274-eaeac6b53751" />

Here is what I mean by crack/gap:
<img width="410" alt="crack:gap example" src="https://github.com/user-attachments/assets/3f098c7f-91fb-41b3-b88f-2ad953384d00" />

**Comment by orchetect** [comment]
Are your system sound effects assigned to the Multi-Output device as well as system audio output? If so, does it still happen when you reassign the sound effects to a different device completely unrelated to the Multi-Output?

**Comment by luismrguimaraes** [comment]
I'm unsure what you mean by "system sound effects".

For example, in Reaper I use this (with the aggregate device I shared above)
<img width="517" alt="image" src="https://github.com/user-attachments/assets/dcd61088-6146-4805-9391-747d86e52ab2" />
In the AudioPluginHost, this is the setting:
<img width="548" alt="image" src="https://github.com/user-attachments/assets/ed00e9e5-d6a5-4bde-ad85-c5c76f559506" />
In Audacity:
<img width="479" alt="image" src="https://github.com/user-attachments/assets/4786514f-de49-4309-8bd2-0d4ce1d33cf1" />

Then, this is what each device settings look like:
<img width="1460" alt="image" src="https://github.com/user-attachments/assets/cea234bb-ef6a-43f5-85aa-1f9578fb47e8" />
<img width="1467" alt="image" src="https://github.com/user-attachments/assets/cc2086a7-cf0e-4d4a-a4e1-1012c4341fe5" />

I'm not using Multi-Output devices, but I do have two created the same way I have two aggregate devices with blackhole:
<img width="344" alt="image" src="https://github.com/user-attachments/assets/f306842a-64d0-4f94-ad27-b4e73b3b1ebc" />

**Comment by devinroth** [comment]
Let me be more specific. There are known issues with drift correction with Apple’s multi-outputs/aggregates. I want you to be able to reproduce this issue without using them.

What is outputting audio to BlackHole?

Devin

**Comment by luismrguimaraes** [comment]
I'm setting the device output to BlackHole
<img width="340" alt="image" src="https://github.com/user-attachments/assets/7a760c1c-44b9-4537-81b0-c49d8c08e5ac" />

and have now tried recording with Reaper with its settings like this:
<img width="523" alt="image" src="https://github.com/user-attachments/assets/4575b839-13c1-43d0-875a-592aa1a0c1d4" />

but am still getting the cracks/gaps
<img width="247" alt="image" src="https://github.com/user-attachments/assets/c545f8e3-0918-4b08-85d0-b0e1347a800d" />

Should I also try deleting the aggregate devices and multi-output that include BlackHole?

**Comment by orchetect** [comment]
A sample buffer ("block size") of 64 is really low and is a frequent cause of dropouts.

Try setting it to 128 or even 256 and see if the dropouts still occur.

**Comment by luismrguimaraes** [comment]
Still happening with 256. I've tested with audio from Spotify, Youtube and VLC.

**Comment by devinroth** [comment]
Okay.

One more thing before I attempt to reproduce. Restart your computer and try again.

Devin

**Comment by luismrguimaraes** [comment]
After restarting, it is not happening in Audacity. It still happens in Reaper (with the aggregate device) and in the AudioPluginHost (with each device selected separately for input and output) but is less frequent the more I increase the buffer size. 1024 samples seems to be the best value with very rare crack/gap occurrences.

I had actually not tried 512 or 1024 buffer sizes before, but I'm not sure if this should be considered a solution to the problem. I'm noticing that what I'm doing to trigger the cracks causes small CPU spikes, but because they are so small (from ~4% to ~8%) I wouldn't expect them to interfere with the audio because they are not high priority (from a user perspective) but they are likely getting priority over BlackHole.

**Comment by luismrguimaraes** [comment]
Actually, if the buffer length is small enough (tested with 5ms and 10ms), it also happens in Audacity.

**Comment by devinroth** [comment]
Are you getting any overload errors in console?

A little background here, BlackHole does not add any additional latency and relies on the buffers of the applications to arrive at the required time so there’s enough internal processing time to do any copy operations. If either application is writing the audio buffer late or if the callback is earlier there will be a click. This can be solved internally by creating a safety offset inside the driver but that is essentially the same as increasing the buffer size of the applications.

The click is clearly being caused by an OS interruption of the real-time audio through. Something that should never happen but I’ve found that many audio developers are “lazy” and instead of ensuring 100% safe operations they cut corners. The correct way to do this is to ensure an application is never interrupted by OS but some libraries are poorly written and are integrated all over the place. That means it ends up being “BlackHole’s fault” and we have to compensate for the poor decisions of other developers.

Adding a small safety offset in BlackHole should solve the problem. But is that really the solution?

Devin

**Comment by devinroth** [comment]
I’m getting all philosophical about the correct way to do things but users be like “I don’t care, I just want it to work”.

Devin

**Comment by luismrguimaraes** [comment]
Thanks for the replies. 

In Console, when the cracks happen, if I search `blackhole` I see this:
<img width="2347" alt="image" src="https://github.com/user-attachments/assets/24e3e91a-0145-41c9-8c4f-51f4bea891fb" />

And I see several messages like this one if I search `audio`:
<img width="1542" alt="image" src="https://github.com/user-attachments/assets/a45eb18d-593f-4069-b346-680c737523fc" />

For setting the safety offset, do I need to do [this](https://github.com/ExistentialAudio/BlackHole/wiki/Adjust-Driver-Latency)? If not, how can I do it?

**Comment by devinroth** [comment]
lol. I don’t even know what that means.

Anything when just searching for coreaudio?

Devin

**Comment by devinroth** [comment]
Sorry I didn’t notice the 2nd screenshot.

So yes as suspected this is an audio overload issue. The solution is a larger buffer size. Or if you want the driver to handle this, increase the safety offset.

Devin

**Comment by luismrguimaraes** [comment]
Thanks Devin, but you also didn't notice my question hehe :) . Is [this](https://github.com/ExistentialAudio/BlackHole/wiki/Adjust-Driver-Latency) was the way to increase the safety offset?

**Comment by devinroth** [comment]
That should work. Try it out and see what happens.

Devin

**Comment by ll14m4n** [comment]
I have same or similar issue. 
I use latest macOS Sequoia 15.2 and Audio Hijack with some EQ and Compression, input is my mic, output is blackhole.
After disabling all effects and setting buffer to max value in Audio Hijack, it still sound like buffer is too small, i.e. small short interrupts in audio stream.
I think it started after I upgraded blackhole from v0.5 to v0.6 via homebrew.

Could you please share a link to the older version 0.5? I can't find it anywhere.

**Comment by Indra-poudel** [comment]
@luismrguimaraes 
Did that help you overcome cracks.

I route the system audio to Blackhole 2ch and capture it and write to file to verify. 
There is also this crack sound in between.

cc @devinroth

**Comment by luismrguimaraes** [comment]
Unfortunately I haven't manage to find the time to try it yet. I do plan on doing so eventually.

**Comment by luismrguimaraes** [comment]
The issue seems worse when the RAM usage is higher. I hadn't restarted my mac in a while and was getting a lot of cracks, even with high buffer sizes (~2048 in Reaper) (and also with the locally built driver with a safety offset of 512). I've now restarted, and I'm able to get rid of the cracks with a buffer size of 512 in Reaper and 40ms of latency in Audacity (its default). I had actually noticed in the past, before using blackhole, that I heard similar cracks when the RAM usage was >14GB which seems like a Ventura or core audio bug.

After filling the RAM usage a bit more (now ~12GB/16GB), and building with a `kLatency_Frame_Size` of 32, I'm still recording cracks in Audacity when I set the latency to 5ms:

<img width="134" alt="Image" src="https://github.com/user-attachments/assets/94c61c84-46df-4b67-aafc-5b259889ee1c" />

with these settings:

<img width="456" alt="Image" src="https://github.com/user-attachments/assets/1431f73a-f8cd-448d-bfb1-dbbd67ca25ec" />

I'm confident I'm using the local blackhole and not the previously installed one because 1. I uninstalled the previous blackhole-2ch and 2. I'm seeing a different icon:

<img width="338" alt="Image" src="https://github.com/user-attachments/assets/c2724d03-6ba7-4b89-9f26-7b4ac0bfcfbc" />

I'm not using the aggregate device in Audacity, but I am in Reaper, which allows me to hear the cracks in real time. Whenever I set similar latency values, the behaviour seems consistent between Reaper (with the aggregate device) and Audacity (without it).

I do still have the aggregate devices setup. Could the drift correction be interfering even though I'm not directly using the aggregate devices?

<img width="1007" alt="Image" src="https://github.com/user-attachments/assets/159d5d78-3689-4cc6-9b23-ee2cf4f5f5d4" />

**Comment by luismrguimaraes** [comment]
So, just to note, I cannot recreate this issue with just the system audio, I'm only able to recreate the cracks when I use blackhole.

I do get overload messages in console, but only when using blackhole (crack "trigger" as at 22:24:59):

<img width="1582" alt="Image" src="https://github.com/user-attachments/assets/46cc25b2-0e63-4e2f-af5c-0bf699b8e738" />
more logs:
<img width="1582" alt="Image" src="https://github.com/user-attachments/assets/7b23fdfe-97a5-46c4-a4c0-e2fa01b58a07" />

Without blackhole, I get this (crack "trigger" was at 22:26:39):

<img width="1582" alt="Image" src="https://github.com/user-attachments/assets/61cb060d-a50d-45d5-82ed-fc0b7d893886" />

I pushed the RAM usage way up for these tests:

<img width="1140" alt="Image" src="https://github.com/user-attachments/assets/13d73cea-751f-4835-971d-f160163fa893" />

**Comment by devinroth** [comment]
This might be the best behaviour given the situation. If there is in fact an overload, then we don't have much of a choice. 

But having a larger safety offset should solve the problem. And it's not. Correct?

**Comment by luismrguimaraes** [comment]
I'm currently using a safety offset of 32 samples.

**Comment by luismrguimaraes** [comment]
I've tested 512 before when RAM usage was ~14GB, and it didn't stop the cracks. I can try a **really** high safety offset and see if I get cracks even at high RAM usage.

**Comment by luismrguimaraes** [comment]
Also, does it make sense that the overloads only happen when using the blackhole driver?

**Comment by devinroth** [comment]
Can you reproduce on Sequoia?

Devin

**Comment by devinroth** [comment]
Yes. Since there is no room for error BlackHole simply can’t wait for late audio.

The built-in devices have high buffer sizes and safely offsets to prevent this type of stuff.

Devin

**Comment by luismrguimaraes** [comment]
> Can you reproduce on Sequoia?
> 
> Devin
> […](#)

I can try tomorrow.


> Yes. Since there is no room for error BlackHole simply can’t wait for late audio.
> 
> The built-in devices have high buffer sizes and safely offsets to prevent this type of stuff.
> 
> Devin
> […](#)

Hmm, interesting. But that means that just increasing the safety offset would fix it. I'm now tempted to try higher buffer sizes than 512.

**Comment by devinroth** [comment]
Yes in theory. Assuming this is a result of either the application taking too long or perhaps a macOS bug.

Classic “it’s not my fault”. But let’s find out.

Devin

**Comment by luismrguimaraes** [comment]
I've just tested with an aggregate device on Sequoia 15.3.1 and cannot recreate the cracks.

**Comment by devinroth** [comment]
Very interesting. This points to a macOS bug.

For clarity, did you test on Ventura without an aggregate?

Devin

**Comment by luismrguimaraes** [comment]
I did. As I mentioned [here](https://github.com/ExistentialAudio/BlackHole/issues/824#issuecomment-2697369684), I used Audacity without the Aggregate Device. The Aggregate Device still existed, but Audacity allows me to choose input and output devices separately.

**Comment by devinroth** [comment]
Very strange.

This definitely points to a MacOS bug that got fixed in Sequoia.

If it’s important to find a solution to this I suggest playing around with the latency / safety offset in the driver.

Devin

**Comment by Omega3komma3** [comment]
Had a similar issue where the BlackHole Output only started crackling when I fed my mic input into it but never with plugins that played back a file. I recompiled the Driver with a latency of 512 and now it works perfectly fine. 

Edit: I am on MacOS 26.3.1 (a)
